README.md
# WebP file format plug-in for Photoshop

Current plug-in version: WebPShop 0.4.3

WebPShop is a Photoshop module for opening and saving WebP images, including
animations.

**Important note: [Photoshop 23.2 and newer natively supports WebP](https://helpx.adobe.com/photoshop/kb/support-webp-image-format.html).**
However some features such as preview at encoding and animations are missing.
WebPShop can still be installed and used for these use cases and/or for
Photoshop 23.1 and below.

Please look at the files LICENSE and CONTRIBUTING in the "docs" folder before
using the contents of this repository or contributing.

## Installation

Download the binary at https://github.com/webmproject/WebPShop/releases. \
Direct link for Windows x64:
https://github.com/webmproject/WebPShop/releases/download/v0.4.3/WebPShop_0_4_3_Win_x64.8bi \
Direct link for MacOS (extract the ZIP archive afterwise):
https://github.com/webmproject/WebPShop/releases/download/v0.4.2/WebPShop_0_4_2_Mac_Universal.zip \
Move the plug-in (the .8bi binary for Windows or the .plugin folder for MacOS)
to the Photoshop plug-in directory
(`C:\Program Files\Common Files\Adobe\Plug-Ins\CC` for Windows,
`/Library/Application Support/Adobe/Plug-Ins/CC` for Mac). Run Photoshop.

On macOS 10.15+, the prompt "WebPShop.plugin cannot be opened because
the developer cannot be verified" can be bypassed by running the following
in Terminal (Finder > Applications > Utilities):

```
sudo xattr -r -d com.apple.quarantine /Library/Application\ Support/Adobe/Plug-Ins/CC/WebPShop.plugin
```

## Features

*   `Open`, `Open As` menu commands can be used to read .webp files.
*   `Save a Copy...` menu command can be used to write .webp files. Encoding
    parameters can be tuned through the UI.

![WebPShop encoding settings - Windows](docs/webpshop_enc_ui_windows.webp)

Photoshop 23.2 and above has partial native WebP support. It will appear as
`WebP (*.WEBP)` in the "Save as type:" drop-down list at encoding. WebPShop will
still appear as `WebPShop (*.WEBP, *.WEBP)`. Use the latter. \
WebPShop can also be used at decoding through the "Open as..." menu.

## Encoding settings

For information, the quality slider maps the following ranges to their internal
WebP counterparts (see SetWebPConfig() in WebPShopEncodeUtils.cpp):

    | Quality slider value   -> | 0    ...    97 | 98         99 |    100   |
    |---------------------------|----------------|---------------|----------|
    | WebP encoding settings -> | Lossy, quality | Near-lossless | Lossless |
    |                           | 0    ...   100 | 60         80 |          |

The radio buttons offer several levels of compression effort:

    | Label   | WebP speed setting  | Sharp YUV    | WebP "quality" setting |
    |         |                     | (lossy only) | (except for lossy)     |
    |---------|---------------------|--------------|------------------------|
    | Fastest |          1          |      No      |            0           |
    | Default |          4          |      No      |           75           |
    | Slowest |          6          |      Yes     |          100           |

## Limitations

*   Only English is currently supported.
*   Only "RGB Color" image mode is currently supported.
*   16 and 32 bits/channels are downscaled to 8 bits/channels before the WebP
    encoding because WebP only supports 8-bit internally.
    Exports from 32-bit documents should include the color profile in the WebP
    encoding settings, otherwise they might appear darker than expected.
*   WebP images cannot exceed 16383 x 16383 pixels.
*   The Timeline data is not used; thus animations rely on layers for defining
    frames (set duration as "(123 ms)" in each layer's name), and they need to
    be rasterized before saving.
*   On some images, lossless compression might produce smaller file sizes than
    lossy. That's why the quality slider is not linear. The same problem exists
    with the radio buttons controlling the compression effort.
*   The color profile is not applied to the Preview image on macOS, regardless
    of the related checkbox state.
*   This plug-in does not extend `Export As` neither `Save for Web`.
*   Encoding and decoding are done in a single pass. It is not currently
    possible to cancel such actions, and it might take some time on big images.
*   Only the latest Photoshop release is supported.

## Troubleshooting

If the plug-in is not detected or does not behave as expected, the steps below
might help:

*   Update Photoshop to the latest version.
*   Double-check that the plug-in binaries match the Operating System and the
    architecture.
*   The plug-in should be listed in the "Help > About Plugins" submenu if it is
    found by Photoshop.
*   If it is undetected, disable any antivirus program or allow the plug-in
    execution (including in MacOS and Windows built-in protections).
*   If it is still undetected, try each of these folders. Windows paths:

        C:\Program Files\Common Files\Adobe\Plug-Ins\CC
        C:\Program Files\Common Files\Adobe\Plug-Ins\CC\File Formats
        C:\Program Files\Adobe\Adobe Photoshop 2022\Plug-ins

    MacOS paths:

        /Library/Application Support/Adobe/Plug-Ins/CC
		Applications/Adobe Photoshop/Plug-ins/

*   If it is still undetected, remove all plug-ins from all folders and copy
    WebPShop in only one of these folders, in case there is a plug-in conflict.
    Restart the computer and/or Photoshop.

If the issue still occurs, check https://github.com/webmproject/WebPShop/issues
to see if it is already mentioned or open a new bug report otherwise.

## Software architecture

The `common` folder contains the following:

*   `WebPShop.h` is the main header, containing most functions.
*   `WebPShop.cpp` contains the plug-in entry point (called by host).
*   `WebPShop.r` and `WebPShopTerminology.h` represent the plug-in properties.
*   Functions in `WebPShopSelector*` are called in `WebPShop.cpp`.
*   `WebPShop*Utils.cpp` are helper functions.
*   `WebPShopScripting.cpp` is mostly used for automation.
*   `WebPShopUI*` display the encoding parameters window and the About box.

The `win` folder contains a Visual Studio solution and project, alongside with
`WebPShop.rc` which is the encoding parameters window layout and About box.

The `mac` folder contains an XCode project. `WebPShopUIDialog_mac.h` and `.mm`
describe the UI layout, while `WebPShopUI_mac.mm` handles the window events.

## Build

Current libwebp version: WebP 1.2.2

Use Microsoft Visual Studio (2019 and above) for Windows and XCode for Mac.

*   Download the latest Adobe Photoshop Plug-In and Connection SDK at
    https://console.adobe.io/downloads/ps,
*   Put the contents of this repository in a "webpshop" folder located at
    `adobe_photoshop_sdk_[version]/pluginsdk/samplecode/format`,
*   Download the latest WebP binaries at
    https://developers.google.com/speed/webp/docs/precompiled or build them,
*   Add `path/to/webp/includes` and `path/to/webp/includes/src` as Additional
    Include Directories to the WebPShop project [1],
*   Add webp, webpdemux, webpmux libraries as Additional Dependencies to the
    WebPShop project [1],
*   Build with the same architecture and configuration as your Photoshop
    installation and the WebP binaries (x64 or arm64, Debug/Release),
*   By example for Windows, it should output the plug-in file `WebPShop.8bi` in
    `adobe_photoshop_sdk_[version]/pluginsdk/samplecode/Output/Win/x64/Release`.

[1] By default the XCode project includes and links to the
`libwebp-[version]-mac-[version]` folder in the `webpshop` directory. The VS
project expects `libwebp-[version]-windows-x64` (or `-arm64`).


docs/NEWS.md
-   2018/10/31: v0.1.0

    *   Initial release of WebPShop, a Photoshop plug-in for reading and writing
        WebP files.
    *   Handles animations too.

-   2019/03/01: v0.1.1

    *   Updated to Adobe Photoshop Plug-In and Connection SDK CC 2019
        (posix i/o for Mac, no handle lock).
    *   Updated to libwebp 1.0.2.
    *   Tested on Windows Server 2016 Datacenter with Photoshop CC 2018
        (v 19.1.7) x64, CC 2019 (v.20.0.3) x64, Release and Debug.

-   2019/03/15: v0.2.0

    *   Now compiles with XCode.
    *   Added Cocoa interface implementation for Mac (encoding settings window
        and About box).
    *   Tested on a MacBook Pro (macOS Mojave 10.14.3) with Photoshop CC 2019
        (v 20.0.3 and 20.0.4).

-   2019/04/24: v0.2.1

    *   Adapted UI quality and compression mappings to WebP encoding settings.
    *   Tested on Windows Server 2016 Datacenter with Photoshop CC 2019
        (v 20.0.4) x64, Release.

-   2020/04/29: v0.3.0

    *   Added the "Loop forever" checkbox for animations.
    *   Added metadata import/export from/to WebP through the "Keep EXIF",
        "Keep XMP" and "Keep Color Profile" checkboxes.
    *   Tested on Windows Server 2016 Datacenter with Photoshop 2020
        (v 21.1.2) x64, Release.

-   2020/06/26: v0.3.1

    *   Fix inaccuracies in WebPShop.r resource file.
    *   Fix PIUtilities.r MacOS ResMerger warning.
    *   Add directions in README.md.

-   2021/02/05: v0.3.2

    *   Updated to Adobe Photoshop Plug-In and Connection SDK 2021 v2 (2021).
    *   Updated to libwebp 1.2.0.
    *   Built with Microsoft Visual Studio Community 2017 15.8.3 and
        Apple XCode 12.4 (12D4e).
    *   Tested on Windows Server 2016 Datacenter with Photoshop 2021
        (v 22.1.1) x64, Release.
    *   Tested on macOS Catalina Version 10.15.7 with Photoshop 2020
        (v 21.2.4) and Photoshop 2021 (v 22.1.1) x64, Release.

-   2021/04/09: v0.3.3

    *   Built with Microsoft Visual Studio Community 2019 16.9.2.
    *   Removed the Windows x86 architecture target.
    *   Added the Windows ARM64 architecture target.
    *   Tested on Windows Server 2016 Datacenter with Photoshop 2021
        (v 22.3.0) x64, Release.

-   2021/06/29: v0.3.4

    *   Fixed the opening of animated WebP images as Smart Objects.

-   2021/09/14: v0.4.0

    *   Updated to libwebp 1.2.1.
    *   Added 16 and 32 bits/channel support.
    *   Display an error prompt instead of hiding WebP export when
        saving an image of another Mode than RGB Color.
    *   Built with Microsoft Visual Studio Community 2019 16.11.0.
    *   Tested on Windows Server 2016 Datacenter with Photoshop 2021
        (v 22.5.1) x64, Release.

-   2021/11/26: v0.4.1

    *   Fix focus issue when closing the WebP encoding settings window on
        Windows.
    *   Built with Microsoft Visual Studio Community 2019 16.11.7.
    *   Tested on Windows Server 2016 Datacenter with Photoshop 2021
        (v 23.0.2) x64, Release.

-   2022/02/21: v0.4.2

    *   Updated to Adobe Photoshop Plug-In and Connection SDK 2021 v1 (2022).
    *   Updated to libwebp 1.2.2.
    *   Added a close button to the About WebP dialog.
    *   Increased the Encoding Settings dialog dimensions (bigger preview).
    *   Built with Microsoft Visual Studio Community 2019 16.11.10.
    *   Tested on Windows Server 2016 Datacenter with Photoshop 2022
        (v 23.2.0) x64, Release.

-   2022/04/01: v0.4.3

    *   Fixed the Preview applying the color profile with no ICC on Windows.
    *   Built with Microsoft Visual Studio Community 2019 16.11.10.
    *   Tested on Windows Server 2016 Datacenter with Photoshop 2022
        (v 23.2.2) x64, Release.


common/WebPShop.r
// # Copyright 2018 Google LLC
// #
// # Licensed under the Apache License, Version 2.0 (the "License");
// # you may not use this file except in compliance with the License.
// # You may obtain a copy of the License at
// #
// #     https://www.apache.org/licenses/LICENSE-2.0
// #
// # Unless required by applicable law or agreed to in writing, software
// # distributed under the License is distributed on an "AS IS" BASIS,
// # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// # See the License for the specific language governing permissions and
// # limitations under the License.
// (Symbol # is used in case this file is wrongly recognized as R language)

//------------------------------------------------------------------------------
// Definitions -- Required by include files.
//------------------------------------------------------------------------------

#define plugInName "WebPShop"
#define plugInCopyrightYear "2020"
#define plugInDescription \
  "A Photoshop plug-in for reading and writing WebP files."

//------------------------------------------------------------------------------
// Definitions -- Required by other resources in this resources file.
//------------------------------------------------------------------------------

// Dictionary (aete) resources:

#define webpshopResourceID 16000  // number seems standard
#define vendorName "Google"
#define plugInAETEComment "webp format module"

#define plugInSuiteID 'gOOG'
#define plugInClassID 'webP'
#define plugInEventID typeNull  // must be this

//------------------------------------------------------------------------------
// Set up included files for Macintosh and Windows.
//------------------------------------------------------------------------------

#include "PIDefines.h"
#include "PIResourceDefines.h"

#if __PIMac__
#include "PIGeneral.r"
#include "PIUtilities.r"
#elif defined(__PIWin__)
#define Rez
#include "PIGeneral.h"
#include "PIUtilities.r"
#endif

#include "PIActions.h"
#include "PITerminology.h"

#include "WebPShopTerminology.h"

//------------------------------------------------------------------------------
// PiPL resource
//------------------------------------------------------------------------------

resource 'PiPL'(webpshopResourceID, plugInName " PiPL", purgeable){
    {Kind{ImageFormat}, Name{plugInName},
     Version{(latestFormatVersion << 16) | latestFormatSubVersion},

     Component{ComponentNumber, plugInName},

#if Macintosh
#if defined(__arm64__)
    CodeMacARM64 { "PluginMain" },
#endif
#if defined(__x86_64__)
    CodeMacIntel64 { "PluginMain" },
#endif
#elif MSWindows
    CodeEntryPointWin64 { "PluginMain" },
#endif

     // This plug-in can read and write via POSIX I/O routines
     SupportsPOSIXIO{},

     // ClassID, eventID, aete ID, uniqueString:
     HasTerminology{plugInClassID, plugInEventID, webpshopResourceID,
                    vendorName " " plugInName},

     // Show or hide the format when saving in a given Image Mode.
     // Always show but display an error in DoOptionsStart() if not RGB.
     SupportedModes{doesSupportBitmap, doesSupportGrayScale,
                    doesSupportIndexedColor, doesSupportRGBColor,
                    doesSupportCMYKColor, doesSupportHSLColor,
                    doesSupportHSBColor, doesSupportMultichannel,
                    doesSupportDuotone, doesSupportLABColor},
     // "EnableInfo and SupportedModes should always match" according to
     // PIGeneral.h File Reference, but SimpleFormat.r just uses "true".
     EnableInfo{"true"},

     // WebP is limited to 16383 in width and height.
     PlugInMaxSize{16383, 16383}, FormatMaxSize{{16383, 16383}},

     FormatMaxChannels{{1, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24}},

     FmtFileType{'WEBP', '8BIM'},
     ReadTypes{{'WEBP', '    '}},
     // FilteredTypes{{'8B1F', '    '}},
     ReadExtensions{{'WEBP'}},
     WriteExtensions{{'WEBP'}},
     FilteredExtensions{{'WEBP'}},
     FormatFlags{fmtSavesImageResources, fmtCanRead, fmtCanWrite,
                 fmtCanWriteIfRead, fmtCanWriteTransparency,
                 fmtCannotCreateThumbnail},
     FormatICCFlags{iccCannotEmbedGray, iccCannotEmbedIndexed,
                    iccCanEmbedRGB, iccCannotEmbedCMYK},

     // XMPRead and XMPWrite properties would enable formatSelectorXMPRead
     // and formatSelectorXMPWrite which represent another way of handling
     // XMP metadata import/export. Currently it is done as EXIF: through
     // handle allocation and setPropertyProc().

     // No way of enabling the "ICC Profile" checkbox in the "Save As" window
     // was found. Currently it is configurable in the "WebPShop" window among
     // other WebP settings and internally handled in a similar way as EXIF and
     // XMP chunks, using the iCCprofileData handle.

     // Layer support is needed for reading animated WebPs into layers
     // (also it would be better to also load it into the timeline).
     FormatLayerSupport{doesSupportFormatLayers},
     // However it is not needed for writing because layer data is directly
     // fetched with CopyAllLayers() thus there is no need to wait for
     // Photoshop and DoWriteLayer*(). Also with FormatLayerSupportReadOnly
     // Ctrl+S defaults to .psd instead of .webp, which is the default
     // behavior for .png etc.
     FormatLayerSupportReadOnly{}}};

//------------------------------------------------------------------------------
// Dictionary (scripting) resource
//------------------------------------------------------------------------------

resource 'aete'(webpshopResourceID, plugInName " dictionary", purgeable){
    1,
    0,
    english,
    roman, /* aete version and language specifiers */
    {
        vendorName,            /* vendor suite name */
        "WebP format plug-in", /* optional description */
        plugInSuiteID,         /* suite ID */
        1,                     /* suite code, must be 1 */
        1,                     /* suite level, must be 1 */
        {},                    /* structure for filters */
        {
            /* non-filter plug-in class here */
            vendorName " WebP format", /* unique class name */
            plugInClassID,          /* class ID, must be unique or Suite ID */
            plugInAETEComment,      /* optional description */
            {                       /* define inheritance */
             "<Inheritance>",       /* must be exactly this */
             keyInherits,           /* must be keyInherits */
             classFormat,           /* parent: Format, Import, Export */
             "parent class format", /* optional description */
             flagsSingleProperty,   /* if properties, list below */

             "Quality",
             keyWriteConfig_quality,
             typeInteger,
             "image quality after compression",
             flagsSingleProperty,

             "Compression",
             keyWriteConfig_compression,
             typeInteger,
             "compression level",
             flagsSingleProperty,

             "Keep EXIF",
             keyWriteConfig_keep_exif,
             typeBoolean,
             "keep EXIF metadata",
             flagsSingleProperty,

             "Keep XMP",
             keyWriteConfig_keep_xmp,
             typeBoolean,
             "keep XMP metadata",
             flagsSingleProperty,

             "Keep Color Profile",
             keyWriteConfig_keep_color_profile,
             typeBoolean,
             "keep color profile",
             flagsSingleProperty,

             "Infinite Loop",
             keyWriteConfig_loop_forever,
             typeBoolean,
             "loop the animation forever",
             flagsSingleProperty,

             "Using POSIX I/O",
             keyUsePOSIX,
             typeBoolean,
             "use POSIX file i/o",
             flagsSingleProperty},
            {}, /* elements (not supported) */
                /* class descriptions */
        },
        {}, /* comparison ops (not supported) */
        {}  /* any enumerations */
    }};

//------------------------------------------------------------------------------
// History resource
//------------------------------------------------------------------------------

resource 'STR '(kHistoryEntry, "History", purgeable){
  plugInName ": ref num=^0."
};

//------------------------------------------------------------------------------

// end WebPShop.r
